airsonic-main/src/main/java/org/airsonic/player/service/MediaScannerService.java

```diff
@@ -191,7 +188,8 @@ private void doScanLibrary() {
             // Maps from artist name to album count.
             Map<String, AtomicInteger> albumCount = new ConcurrentHashMap<>();
             Map<String, Artist> artists = new ConcurrentHashMap<>();
-            Map<MediaFile, Album> albums = new ConcurrentHashMap<>();
+            Map<String, Album> albums = new ConcurrentHashMap<>();
```

Comment:
What was the impetus for changing this from using `MediaFile`?

Reply:
That's a good question. I'll try explaining from the beginning.

- The reason these maps were put in here in the first place is to accumulate data "in memory" so we can put/commit everything to the db together at the END of the media scanning process. Going back and forth to the DB DURING the media scanning process seemed like an expensive proposition to me, especially sometimes for the same data, given the latencies involved with a roundtrip to the DB.
- My reasoning formed when I saw this pattern happening:
  - Encounter a file
  - Fetch or create a row for it from the db
  - change some attribute for it in the db
  - a different method: fetch or create a row for it from the db (which has no idea about a different method doing the same thing, possibly in parallel)
  - change some attribute for it in the db
  - and so on
- I thought going to the db back and forth for every operation (sometimes changing the same thing at times) DURING scanning was an expensive proposition, and it might be helpful if we had an "in-memory db" that we just add to or update or take from while scanning and then commit the final product to the db in whole at the end.
- So, imagine we're scanning Artist1's Album1, and it has Song1, Song2, and Song3. And let's say this artist/album is the first time ever encoutered by the media scanner
  - Previously:
    - Encounter Song1: Create Artist1 in the DB, fetch that row, update its timestamp and recommit to DB. Create Album1 in the DB, update its timestamp, recommit. Create Song1.
    - Encounter Song2: Create Artist1 in the DB (already there), update its timestamp. Create Album1 (already there), update its timestamp. Create Song2.
    - Encounter Song3: Create Artist1 in the DB (already there), update its timestamp. Create Album1 (already there), update its timestamp. Create Song3.
  - Now:
    - Encounter Song1: Create Artist1 in the Map if not exists (doesn't), fetch that entry, update its timestamp. Create Album1 in the Map, update its timestamp. Create Song1.
    - Encounter Song2: Create Artist1 in the Map if not exists (it does, so its a no-op), update its timestamp. Create Album1 in the Map (noop), update its timestamp. Create Song2.
    - Encounter Song3: Create Artist1 in the Map if not exists (it does, so its a no-op), update its timestamp. Create Album1 in the Map (noop), update its timestamp. Create Song3.
    - At the end, commit the final Artist1 result, the final Album1 result to the DB. We've successfully minimized the churn DURING processing to in-memory structures instead of going back and forth with the DB, and the DB only sees the final product. We pay the penalty for going to the DB only once instead of of like 9 times every time we need to update (or create) something.
- So now we need to hold the data in memory, we need a key to address it with, during the in-processing and at the end during committing. This should be whatever the primary key in the db was (artist's name etc), or some other unique key they were using to retrieve and update the data in memory. For albums I made a mistake using a MediaFile as the key. My thought process was that it would create a MediaFile for an album, and then retrieve and update it on subsequent accesses, but what I forgot was that each song has a different MediaFile. So here's what ends up happening:
  - Album1 folder encountered. MediaFile1 created for Album1. Check map for album under MediaFile1 key, none found. Create album for Album1 and store in the Map with MediaFile1 as key.
  - Album1/Song1 encountered. MediaFile2 for Song1 created. Check map for album under MediaFile2 key, none found. Create album for Album1 and store in the Map with MediaFile2 as key.
  - Continue for all other songs. At the end, our map contains the same album split across (lets say 10 songs + 1 album), 11 keys, each with partial info. Instead of aggregating into one album, each song doesn't find the album and so creates one. And when we finally go to commit to the db, the DB starts throwing unique key violations. It'll create the first key's album, then will encounter the same album under a different key, and will say, I already have this album, cannot commit, and so forth.
- Thus this is the solution, the String key of the album map reflects the primary key of the album in the db (artist name/album name composite), and is generated here: https://github.com/airsonic-advanced/airsonic-advanced/pull/20/files#diff-44b0964f8fb27f359ad54d7377ef98b7R333-R334

Reply:
Using a composite key of artist,album for albums doesn't seem like a good thing to me, as it will combine anything with matching artist and track names into the same "album".  This could be problematic in cases like someone has multiple rips of the same album in different formats and these would all be crammed together and make them painful to use.  Or if a bunch of different untagged albums are are labeled with generic "Unknown" values, these would also all end up mixed together in one hard to use "album".  Also, it seems like that might require additional protection code if the media files in the same "album" are spread across different media folders.

It would seem safer and better (to me) to use the deepest directory path (relative to the current top-level media folder) that contains the media file as the album primary key.  I specify that it's a relative path so that if media folders are moved but the structure below them is maintained, the database will find the moved files in the same "place", so the files won't need anything special to preserve ratings, playcounts, or other useful data.

Reply:
Hmm, I think you're conflating MediaFiles (a separate table) with Albums. Whether I move the media files or not should have no bearing on the albums. And perhaps points to the existing (I didn't implement it) underlying design decision here (paths don't matter to albums).

Maybe it wasn't clear what I wrote. Whether it's a good idea or not doesn't matter, we're mandated by the current set up of the DB. You _have_ to use the composite key of (artist,album) because that's what the _unique key constraint_ of the album table in the database is (_NOT_ the path). Unless I change the table structure itself, whatever I use as a key for the in-memory map doesn't matter because once I try persisting in the db, that's the only determining factor.

Let's consider your scenario of album rips in different formats. Let's adopt your solution and say I use different keys for each in the map. At the end of the scan, this is how it's aggregated in the map (which is how you'd like it to be):
key1 -> Album1 (formatA) (Artist1), in some path
key2 -> Album1 (formatB) (Artist1), in a different path

Now lets try committing it to db:
- key1 -> Persists a row in the table with Artist1, Album1 (doesn't matter what the path is, though that's persisted too)
- key2 -> TRIES to persist a row in the table with the same Artist1, Album1, but immediately throws a unique key violation because a row with that composite key already exists. It doesn't matter that another column (the path) is different in this key's row

See the issue? For reference here's the key statement from the DDL of the albums table:
```
CREATE UNIQUE INDEX IDX_ALBUM_ARTIST_NAME ON ALBUM (ARTIST,NAME);
```
There is nothing, not even an index on the `PATH` column. So creating keys based on the path will undoubtedly lead to unique key violations when you try and persist in the database, because two paths with the same album name and artist will undoubtedly conflict in the DB.

Now of course we can redesign the table itself, but that'll be a different undertaking. I'm trying to parallelize the intake, not change the table structure itself. Currently the db as it's set up aggregates into the same albums no matter the path (and maybe that's by design). That's exactly also what the current airsonic installations do. Have you tried seeing how multiple rips of the same album look in your db table? Or how the unknowns are accumulated? They're all aggregated together.

Additionally, there is a flaw with using paths as a unique key: it implies a certain folder structure, which not everyone has. Songs are not always under an albums folder (some of my installations actually have all music under a "Music" folder barebones), I've seen others that rely on a genre-based segregation, and so on. I would not want airsonic to aggregate it all under the Music album. I would want it to read the metadata and aggregate it based on the metadata. But of course that has a different problem as you point out. Maybe adding a format key as part of the uniqueness constraint could be a solution, but then there is a further matter of a rip with half the songs in one format and the other in another...

So in short, redesigning albums needs to be handled with care, but isn't the concern of this PR. It just retains the existing infrastructure.